man/osm2pgsql.1
.TH "OSM2PGSQL" "1" "2.3.1" "" ""
.SH NAME
osm2pgsql \- OpenStreetMap data to PostgreSQL converter
.SH SYNOPSIS
\f[B]osm2pgsql\f[R] [\f[I]OPTIONS\f[R]] OSM\-FILE\&...
.SH DESCRIPTION
\f[B]osm2pgsql\f[R] imports OpenStreetMap data into a PostgreSQL/PostGIS
database.
It is an essential part of many rendering toolchains, the Nominatim
geocoder and other applications processing OSM data.
.PP
\f[B]osm2pgsql\f[R] can run in either \[lq]create\[rq] mode (the
default) or in \[lq]append\[rq] mode (option \f[B]\-a, \-\-append\f[R]).
.PP
In \[lq]create\[rq] mode osm2pgsql will create the database tables
required by the configuration and import the OSM file(s) specified on
the command line into those tables.
Note that you also have to use the \f[B]\-s, \-\-slim\f[R] option if you
want your database to be updatable.
.PP
In \[lq]append\[rq] mode osm2pgsql will update the database tables with
the data from OSM change files specified on the command line.
.PP
This man page can only cover some of the basics and describe the command
line options.
See the \c
.UR https://osm2pgsql.org/doc/manual.html
Osm2pgsql Manual
.UE \c
\ for more information.
.SH OPTIONS
This program follows the usual GNU command line syntax, with long
options starting with two dashes (\f[CR]\-\-\f[R]).
Mandatory arguments to long options are mandatory for short options too.
.SH MAIN OPTIONS
.TP
\-a, \-\-append
Run in append mode.
Adds the OSM change file into the database without removing existing
data.
.TP
\-c, \-\-create
Run in create mode.
This is the default if \f[B]\-a, \-\-append\f[R] is not specified.
Removes existing data from the database tables!
.SH HELP/VERSION OPTIONS
.TP
\-h, \-\-help
Print help.
.TP
\-V, \-\-version
Print osm2pgsql version.
.SH LOGGING OPTIONS
.TP
\-\-log\-level=LEVEL
Set log level (`debug', `info' (default), `warn', or `error').
.TP
\-\-log\-progress=VALUE
Enable (\f[CR]true\f[R]) or disable (\f[CR]false\f[R]) progress logging.
Setting this to \f[CR]auto\f[R] will enable progress logging on the
console and disable it if the output is redirected to a file.
Default: true.
.TP
\-\-log\-sql
Enable logging of SQL commands for debugging.
.TP
\-\-log\-sql\-data
Enable logging of all data added to the database.
This will write out a huge amount of data!
For debugging.
.TP
\-v, \-\-verbose
Same as \f[CR]\-\-log\-level=debug\f[R].
.SH DATABASE OPTIONS
.TP
\-d, \-\-database=NAME
The name of the PostgreSQL database to connect to.
If this parameter contains an \f[CR]=\f[R] sign or starts with a valid
URI prefix (\f[CR]postgresql://\f[R] or \f[CR]postgres://\f[R]), it is
treated as a conninfo string.
See the PostgreSQL manual for details.
.TP
\-U, \-\-username=NAME, \-\-user=NAME
Postgresql user name.
.TP
\-W, \-\-password
Force password prompt.
.TP
\-H, \-\-host=HOSTNAME
Database server hostname or unix domain socket location.
.TP
\-P, \-\-port=PORT
Database server port.
.TP
\-\-schema=SCHEMA
Default for various schema settings throughout osm2pgsql (default:
\f[CR]public\f[R]).
The schema must exist in the database and be writable by the database
user.
.SH INPUT OPTIONS
.TP
\-r, \-\-input\-reader=FORMAT
Select format of the input file.
Available choices are \f[B]auto\f[R] (default) for autodetecting the
format, \f[B]xml\f[R] for OSM XML format files, \f[B]o5m\f[R] for o5m
formatted files and \f[B]pbf\f[R] for OSM PBF binary format.
.TP
\-b, \-\-bbox=MINLON,MINLAT,MAXLON,MAXLAT
Apply a bounding box filter on the imported data.
Example: \f[B]\-\-bbox\f[R] \f[B]\-0.5,51.25,0.5,51.75\f[R]
.SH MIDDLE OPTIONS
.TP
\-i, \-\-tablespace\-index=TABLESPC
Store all indexes in the PostgreSQL tablespace \f[CR]TABLESPC\f[R].
This option also affects the tables created by the pgsql output.
This option is deprecated.
Use the \-\-tablespace\-slim\-index and/or \-\-tablespace\-main\-index
options instead.
.TP
\-\-tablespace\-slim\-data=TABLESPC
Store the slim mode tables in the given tablespace.
.TP
\-\-tablespace\-slim\-index=TABLESPC
Store the indexes of the slim mode tables in the given tablespace.
.TP
\-p, \-\-prefix=PREFIX
Prefix for table names (default: \f[CR]planet_osm\f[R]).
.TP
\-s, \-\-slim
Store temporary data in the database.
Without this mode, all temporary data is stored in RAM and if you do not
have enough the import will not work successfully.
With slim mode, you should be able to import the data even on a system
with limited RAM, although if you do not have enough RAM to cache at
least all of the nodes, the time to import the data will likely be
greatly increased.
.TP
\-\-drop
Drop the slim mode tables from the database and the flat node file once
the import is complete.
This can greatly reduce the size of the database, as the slim mode
tables typically are the same size, if not slightly bigger than the main
tables.
It does not, however, reduce the maximum spike of disk usage during
import.
It can furthermore increase the import speed, as no indexes need to be
created for the slim mode tables, which (depending on hardware) can
nearly halve import time.
Slim mode tables however have to be persistent if you want to be able to
update your database, as these tables are needed for diff processing.
.TP
\-C, \-\-cache=NUM
Only for slim mode: Use up to \f[B]NUM\f[R] MB of RAM for caching nodes.
Giving osm2pgsql sufficient cache to store all imported nodes typically
greatly increases the speed of the import.
Each cached node requires 8 bytes of cache, plus about 10% \- 30%
overhead.
As a rule of thumb, give a bit more than the size of the import file in
PBF format.
If the RAM is not big enough, use about 75% of memory.
Make sure to leave enough RAM for PostgreSQL.
It needs at least the amount of \f[CR]shared_buffers\f[R] given in its
configuration.
Defaults to 800.
.TP
\-x, \-\-extra\-attributes
Include attributes of each object in the middle tables and make them
available to the outputs.
Attributes are: user name, user id, changeset id, timestamp and version.
.TP
\-F, \-\-flat\-nodes=FILENAME
Use a file on disk to store node locations instead of storing them in
memory (in non\-slim mode) or in the database (in slim mode).
This is much more efficient than storing the data in the database.
Storing the node information for the full planet requires more than
500GB in PostgreSQL, the same data is stored in \[lq]only\[rq] 90GB
using the flat\-nodes mode.
This can also increase the speed of applying diff files.
This option activates the flat\-nodes mode and specifies the location of
the database file.
It is a single large file.
This mode is only recommended for full planet imports as it doesn\[cq]t
work well with small imports.
The default is disabled.
The file will stay on disk after import, use \-\-drop to remove it (but
you can\[cq]t do updates then).
.TP
\-\-middle\-schema=SCHEMA
Use PostgreSQL schema SCHEMA for all tables, indexes, and functions in
the middle.
The schema must exist in the database and be writable by the database
user.
By default the schema set with \f[CR]\-\-schema\f[R] is used, or
\f[CR]public\f[R] if that is not set.
.TP
\-\-middle\-with\-nodes
When a flat nodes file is used, nodes are not stored in the database.
Use this option to force storing nodes with tags in the database, too.
.SH OUTPUT OPTIONS
.TP
\-O, \-\-output=OUTPUT
Specifies the output to use.
Currently osm2pgsql supports \f[B]pgsql\f[R], \f[B]flex\f[R], and
\f[B]null\f[R].
\f[B]pgsql\f[R] is the default output still available for backwards
compatibility.
New setups should use the \f[B]flex\f[R] output which allows for a much
more flexible configuration.
The \f[B]null\f[R] output does not write anything and is only useful for
testing or with \f[B]\-\-slim\f[R] for creating slim tables.
.TP
\-S, \-\-style=FILE
The style file.
This specifies how the data is imported into the database, its format
depends on the output.
(For the \f[B]pgsql\f[R] output, the default is
\f[CR]/usr/share/osm2pgsql/default.style\f[R], for other outputs there
is no default.)
.SH PGSQL OUTPUT OPTIONS
.TP
\-\-tablespace\-main\-data=TABLESPC
Store the data tables in the PostgreSQL tablespace \f[CR]TABLESPC\f[R].
.TP
\-\-tablespace\-main\-index=TABLESPC
Store the indexes in the PostgreSQL tablespace \f[CR]TABLESPC\f[R].
.TP
\-\-latlong
Store coordinates in degrees of latitude & longitude.
.TP
\-m, \-\-merc
Store coordinates in Spherical Mercator (Web Mercator, EPSG:3857) (the
default).
.TP
\-E, \-\-proj=SRID
Use projection EPSG:SRID.
.TP
\-p, \-\-prefix=PREFIX
Prefix for table names (default: \f[CR]planet_osm\f[R]).
This option affects the middle as well as the pgsql output table names.
.TP
\-\-tag\-transform\-script=SCRIPT
Specify a Lua script to handle tag filtering and normalisation.
The script contains callback functions for nodes, ways and relations,
which each take a set of tags and returns a transformed, filtered set of
tags which are then written to the database.
.TP
\-x, \-\-extra\-attributes
Include attributes (user name, user id, changeset id, timestamp and
version).
This also requires additional entries in your style file.
.TP
\-k, \-\-hstore
Add tags without column to an additional hstore (key/value) column in
the database tables.
.TP
\-j, \-\-hstore\-all
Add all tags to an additional hstore (key/value) column in the database
tables.
.TP
\-z, \-\-hstore\-column=PREFIX
Add an additional hstore (key/value) column named \f[CR]PREFIX\f[R]
containing all tags that have a key starting with \f[CR]PREFIX\f[R], eg
\f[CR]\[rs]\-\-hstore\-column \[dq]name:\[dq]\f[R] will produce an extra
hstore column that contains all \f[CR]name:xx\f[R] tags.
.TP
\-\-hstore\-match\-only
Only keep objects that have a value in at least one of the non\-hstore
columns.
.TP
\-\-hstore\-add\-index
Create indexes for all hstore columns after import.
.TP
\-G, \-\-multi\-geometry
Normally osm2pgsql splits multi\-part geometries into separate database
rows per part.
A single OSM object can therefore use several rows in the output tables.
With this option, osm2pgsql instead generates multi\-geometry features
in the PostgreSQL tables.
.TP
\-K, \-\-keep\-coastlines
Keep coastline data rather than filtering it out.
By default objects tagged \f[CR]natural=coastline\f[R] will be discarded
based on the assumption that Shapefiles generated by OSMCoastline
(https://osmdata.openstreetmap.de/) will be used for the coastline data.
.TP
\-\-reproject\-area
Compute area column using spherical mercator coordinates even if a
different projection is used for the geometries.
.TP
\-\-output\-pgsql\-schema=SCHEMA
Use PostgreSQL schema SCHEMA for all tables, indexes, and functions in
the pgsql output.
The schema must exist in the database and be writable by the database
user.
By default the schema set with \f[CR]\-\-schema\f[R] is used, or
\f[CR]public\f[R] if that is not set.
.SH EXPIRE OPTIONS
.TP
\-e, \-\-expire\-tiles=[MIN_ZOOM\-]MAX\-ZOOM
Create a tile expiry list.
.TP
\-o, \-\-expire\-output=FILENAME
Output file name for expired tiles list.
.TP
\-\-expire\-bbox\-size=SIZE
Max size for a polygon to expire the whole polygon, not just the
boundary.
.SH ADVANCED OPTIONS
.TP
\-I, \-\-disable\-parallel\-indexing
Disable parallel clustering and index building on all tables, build one
index after the other.
.TP
\-\-number\-processes=THREADS
Specifies the number of parallel threads used for certain operations.
.SH SEE ALSO
.IP \[bu] 2
\c
.UR https://osm2pgsql.org
osm2pgsql website
.UE \c
.IP \[bu] 2
\c
.UR https://osm2pgsql.org/doc/manual.html
osm2pgsql manual
.UE \c
.IP \[bu] 2
\f[B]postgres\f[R](1)
.IP \[bu] 2
\f[B]osmcoastline\f[R](1)


NEWS
== 2010-11-06 Version 0.70.5 ==

* missing libraries are now already detected by 'configure'
  instead of only breaking the build at compile time later

* the existing 'libxml2' and experimental 'primitive' XML
  readers and a new 'pbf' reader for the Protobuf based
  format are now all available in the main binary, 
  selectable using the new -r|--input-reader command line
  option

src/osm2pgsql.cpp
/**
 * SPDX-License-Identifier: GPL-2.0-or-later
 *
 * This file is part of osm2pgsql (https://osm2pgsql.org/).
 *
 * Copyright (C) 2006-2026 by the osm2pgsql developer community.
 * For a full list of authors see the git log.
 */

#include "command-line-parser.hpp"
#include "input.hpp"
#include "logging.hpp"
#include "middle.hpp"
#include "options.hpp"
#include "osmdata.hpp"
#include "output.hpp"
#include "pgsql.hpp"
#include "pgsql-capabilities.hpp"
#include "properties.hpp"
#include "util.hpp"
#include "version.hpp"

#include <osmium/util/memory.hpp>

#include <exception>
#include <filesystem>
#include <memory>
#include <string>
#include <utility>

namespace {

/**
 * Output overall memory usage as debug message.
 *
 * This only works on Linux.
 */
void show_memory_usage()
{
    osmium::MemoryUsage const mem;
    if (mem.peak() != 0) {
        log_debug("Overall memory usage: peak={}MByte current={}MByte",
                  mem.peak(), mem.current());
    }
}

file_info run(options_t const &options, properties_t *properties)
{
    auto const files = prepare_input_files(
        options.input_files, options.input_format, options.append);

    auto thread_pool = std::make_shared<thread_pool_t>(
        options.parallel_indexing ? options.num_procs : 1U);
    log_debug("Started pool with {} threads.", thread_pool->num_threads());

    auto middle = create_middle(thread_pool, options);
    middle->start();

    auto output = output_t::create_output(middle->get_query_instance(),
                                          thread_pool, options, *properties);

    middle->set_requirements(output->get_requirements());

    if (!options.append) {
        properties->init_table();
    }
    properties->store();

    osmdata_t osmdata{middle, output, options};

    // Processing: In this phase the input file(s) are read and parsed,
    // populating some of the tables.
    auto finfo = process_files(files, &osmdata, options.append,
                               get_logger().show_progress());

    show_memory_usage();

    // Process pending ways and relations. Cluster database tables and
    // create indexes.
    osmdata.stop();

    return finfo;
}

void check_db(options_t const &options)
{
    pg_conn_t const db_connection{options.connection_params, "check"};

    init_database_capabilities(db_connection);

    auto const pv = get_postgis_version();
    if (pv.major < 3) {
        throw std::runtime_error{"Need at least PostGIS version 3.0"};
    }

    check_schema(options.dbschema);
    check_schema(options.middle_dbschema);
    check_schema(options.output_dbschema);
}

// This is called in "create" mode to initialize properties.
void set_up_properties(properties_t *properties, options_t const &options)
{
    properties->set_bool("attributes", options.extra_attributes);

    if (options.flat_node_file.empty()) {
        properties->set_string("flat_node_file", "");
    } else {
        properties->set_string(
            "flat_node_file", std::filesystem::absolute(
                                  std::filesystem::path{options.flat_node_file})
                                  .string());
    }

    properties->set_string("prefix", options.prefix);
    properties->set_bool("updatable", options.slim && !options.droptemp);
    properties->set_string("version", get_osm2pgsql_short_version());
    properties->set_int("db_format", options.middle_database_format);
    properties->set_string("output", options.output_backend);

    if (options.style.empty()) {
        properties->set_string("style", "");
    } else {
        properties->set_string(
            "style",
            std::filesystem::absolute(std::filesystem::path{options.style})
                .string());
    }
}

void store_data_properties(properties_t *properties, file_info const &finfo)
{
    if (finfo.last_timestamp.valid()) {
        auto const timestamp = finfo.last_timestamp.to_iso();
        properties->set_string("import_timestamp", timestamp);
        properties->set_string("current_timestamp", timestamp);
    }

    for (std::string const s : {"base_url", "sequence_number", "timestamp"}) {
        auto const value = finfo.header.get("osmosis_replication_" + s);
        if (!value.empty()) {
            properties->set_string("replication_" + s, value);
        }
    }
}

void check_updatable(properties_t const &properties)
{
    if (properties.get_bool("updatable", false)) {
        return;
    }

    throw std::runtime_error{
        "This database is not updatable. To create an"
        " updatable database use --slim (without --drop)."};
}

void check_attributes(properties_t const &properties, options_t *options)
{
    bool const with_attributes = properties.get_bool("attributes", false);

    if (options->extra_attributes) {
        if (!with_attributes) {
            throw std::runtime_error{
                "Can not update with attributes (-x/--extra-attributes)"
                " because original import was without attributes."};
        }
        return;
    }

    if (with_attributes) {
        log_info("Updating with attributes (same as on import).");
        options->extra_attributes = true;
    }
}

void check_and_update_flat_node_file(properties_t *properties,
                                     options_t *options)
{
    auto const flat_node_file_from_import =
        properties->get_string("flat_node_file", "");
    if (options->flat_node_file.empty()) {
        if (flat_node_file_from_import.empty()) {
            log_info("Not using flat node file (same as on import).");
        } else {
            options->flat_node_file = flat_node_file_from_import;
            log_info("Using flat node file '{}' (same as on import).",
                     flat_node_file_from_import);
        }
    } else {
        auto const absolute_path =
            std::filesystem::absolute(
                std::filesystem::path{options->flat_node_file})
                .string();

        if (flat_node_file_from_import.empty()) {
            throw fmt_error("Database was imported without flat node file. Can"
                            " not use flat node file '{}' now.",
                            options->flat_node_file);
        }

        if (absolute_path == flat_node_file_from_import) {
            log_info("Using flat node file '{}' (same as on import).",
                     flat_node_file_from_import);
        } else {
            log_info(
                "Using the flat node file you specified on the command line"
                " ('{}') instead of the one used on import ('{}').",
                absolute_path, flat_node_file_from_import);
            properties->set_string("flat_node_file", absolute_path);
        }
    }
}

void check_prefix(properties_t const &properties, options_t *options)
{
    auto const prefix = properties.get_string("prefix", "planet_osm");
    if (!options->prefix_is_set) {
        log_info("Using prefix '{}' (same as on import).", prefix);
        options->prefix = prefix;
        return;
    }

    if (prefix != options->prefix) {
        throw fmt_error("Different prefix specified on command line ('{}')"
                        " then used on import ('{}').",
                        options->prefix, prefix);
    }
}

void check_db_format(properties_t const &properties, options_t *options)
{
    auto const format = properties.get_int("db_format", -1);

    if (format == 1) {
        throw std::runtime_error{
            "Old database format detected. This version of osm2pgsql can not "
            "read this any more. Downgrade osm2pgsql or reimport database."};
    }

    if (format != 2) {
        throw fmt_error("Unknown db_format '{}' in properties.", format);
    }

    options->middle_database_format = static_cast<uint8_t>(format);
}

void check_output(properties_t const &properties, options_t *options)
{
    auto const output = properties.get_string("output", "pgsql");

    if (options->output_backend.empty()) {
        options->output_backend = output;
        log_info("Using output '{}' (same as on import).", output);
        return;
    }

    if (options->output_backend == output) {
        return;
    }

    throw fmt_error("Different output specified on command line ('{}')"
                    " then used on import ('{}').",
                    options->output_backend, output);
}

void check_and_update_style_file(properties_t *properties, options_t *options)
{
    auto const style_file_from_import = properties->get_string("style", "");

    if (options->style.empty()) {
        log_info("Using style file '{}' (same as on import).",
                 style_file_from_import);
        options->style = style_file_from_import;
        return;
    }

    if (style_file_from_import.empty()) {
        throw std::runtime_error{"Style file from import is empty!?"};
    }

    auto const absolute_path =
        std::filesystem::absolute(std::filesystem::path{options->style})
            .string();

    if (absolute_path == style_file_from_import) {
        log_info("Using style file '{}' (same as on import).",
                 style_file_from_import);
        return;
    }

    log_info("Using the style file you specified on the command line"
             " ('{}') instead of the one used on import ('{}').",
             absolute_path, style_file_from_import);
    properties->set_string("style", absolute_path);
}

// This is called in "append" mode to check that the command line options are
// compatible with the properties stored in the database.
void check_and_update_properties(properties_t *properties, options_t *options)
{
    check_updatable(*properties);
    check_attributes(*properties, options);
    check_and_update_flat_node_file(properties, options);
    check_prefix(*properties, options);
    check_db_format(*properties, options);
    check_output(*properties, options);
    check_and_update_style_file(properties, options);
}

void set_option_defaults(options_t *options)
{
    if (options->output_backend.empty()) {
        options->output_backend = "pgsql";
    }

    if (options->style.empty()) {
        if (options->output_backend == "flex") {
            throw std::runtime_error{"You have to set the config file "
                                     "with the -S|--style option."};
        }
        if (options->output_backend == "pgsql") {
            options->style = DEFAULT_STYLE;
        }
    }
}

} // anonymous namespace

// NOLINTNEXTLINE(bugprone-exception-escape)
int main(int argc, char *argv[])
{
    try {
        auto options = parse_command_line(argc, argv);

        if (options.command == command_t::help) {
            // Already handled inside parse_command_line()
            return 0;
        }

        if (options.command == command_t::version) {
            print_version("osm2pgsql");
            return 0;
        }

        util::timer_t timer_overall;

        check_db(options);

        properties_t properties{options.connection_params,
                                options.middle_dbschema};

        if (options.append) {
            if (!properties.load()) {
                throw std::runtime_error{
                    "Did not find table 'osm2pgsql_properties' in database. "
                    "Database too old? Wrong schema?"};
            }

            check_and_update_properties(&properties, &options);
            properties.store();

            auto const finfo = run(options, &properties);

            if (finfo.last_timestamp.valid()) {
                auto const current_timestamp =
                    properties.get_string("current_timestamp", "");

                if (current_timestamp.empty() ||
                    (finfo.last_timestamp >
                     osmium::Timestamp{current_timestamp})) {
                    properties.set_string("current_timestamp",
                                          finfo.last_timestamp.to_iso());
                }
            }
        } else {
            set_option_defaults(&options);
            set_up_properties(&properties, options);
            auto const finfo = run(options, &properties);
            store_data_properties(&properties, finfo);
        }

        properties.store();

        show_memory_usage();
        log_info("osm2pgsql took {} overall.",
                 util::human_readable_duration(timer_overall.stop()));
    } catch (std::exception const &e) {
        log_error("{}", e.what());
        return 1;
    } catch (...) {
        log_error("Unknown exception.");
        return 1;
    }

    return 0;
}
